enhancements/oc/cli-manager.md

```diff
@@ -0,0 +1,231 @@
+* With `krew index add https://someother-third-party-index` we won't limit users from adding their own index with whatever plugins they want
+* The controller could be optional, since connected environments can use the default `krew` index out-of-the box
+* As `krew` itself is a `kubectl` plugin, it can be invoked using either using `kubectl krew` or `oc krew`
+
+Existing methods of downloading binaries (i.e. the console) will not be affected by this proposal. For the initial implementation, supported plugins will create Plugin CRs. The plugins will be downloaded and installed using `krew`.
+
+### User Stories
+
+#### Story 1
+
+As a user, I want a CLI manager for various CLIs and plugins available for Kubernetes/OpenShift and related services so that I can discover, install, and list them. If `odo` was made available by a cluster-admin, I could install it using:
+* `oc krew install odo`
+
+I could then interact with `odo`:
+* `oc odo --help`
+
+##### Example
+```text
+$ oc krew search odo
+NAME                 DESC                        LATEST       INSTALLED
+-----                -----                       -----        -----
+odo                  OpenShift Developer CLI     1.0          Not Installed
+```
+
+#### Story 2
+
+As owner of a CLI or plugin, I want to publish it to users of the cluster. I need to create a Plugin CR for my tool, or provide the required information about my tool
+to a cluster-admin for the creation of a CR:
+* Name
+* Short description
+* Long description
+* Caveats
+* Homepage
+* Version
+* Platform/architecture
+* The image:tag (and registry credentials if required)
+* The path within the image where the binary for the given platform/architecture can be found
+
+### API Extensions
+
+A new CRD will be created:
+* API: `config.openshift.io/v1`
+* Kind: `Plugin`
+
+### Risks and Mitigations
+
+Distributing undesirable binaries is always a risk. Some mitigations include requiring cluster-admins to maintain the index, and the verification of downloaded
+binaries using SHA256 hashes. Cluster-admins are responsible for publishing only trusted binaries.
+
+## Design Details
+
+Each plugin will provide an image.
+Each plugin is responsible for creating a CR to hold metadata.  The CR will serve to deliver the metadata and description
+of its deliverable binary. The Controller will use CRs to generate an index for the `oc krew search` command, and `oc krew install <name>` will download the binary from the Controller.
+Users will install OpenShift tools that are known compatible with each cluster version through `oc krew`.
+
+A plugin must provide a Plugin CR. The result of this proposal will be:
+* Plugin Custom Resource Definition compatible with `krew`
+* Must work with image registries that require image pull secrets
+* Use `krew` manage plugins made available via CRs
+* A Controller to manage plugins that will serve binaries from images
```

Comment:
this implies the controller will be pulling images, this introduces the possibility of a delta in image pulling behavior between what happens if you create a pod that refs the image, and what this controller does, with respect to:

* use of the cluster scoped image pull secret
* respecting ICSP rules for pulling from mirrors
* use of CAs to talk to the registry
* use of proxies to talk to the registry

this has been a long struggle for builds (one of the main things that does its own image pulling on the cluster).  So if you're going to do this, it needs to be done right and then kept in sync as crio/pod image pulling behavior may continue to change in the future (as it did when ICSP was introduced)

Reply:
@soltysh i don't really see how this was addressed in the design.....   are you going to use a sidecar approach to allow k8s to handle the image pulling for you and avoid the problem, or is there another plan?

any sort of details about how the images will be pulled/extracted seems like a good thing to have in the design details.

Reply:
@deejross ^

Reply:
I have added these points to the Design Details section, as they are important.